First commits to create cyhy-db

felddy opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR contains the initial functionality of the `cyhy-db` module, which provides a object model for accessing the various document collections in a Cyber Hygiene database.

This PR also includes an assortment of unit tests, though more testing will likely be added later.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We have two main motivations here:

1. To migrate from [our legacy data model](https://github.com/cisagov/cyhy-core/blob/develop/cyhy/db/database.py) which was built on outdated Python 2 libraries to a more modern model that uses Python 3 (hello [Beanie](https://beanie-odm.dev/)!)
2. To break the CyHy database model out into its own repository and module thus making it more testable and maintainable.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

All unit tests currently pass in `pytest`.  This code has also been partially tested by its use in the `cyhy-kevsync` module ([initial PR coming soon](https://github.com/cisagov/cyhy-kevsync/tree/first-commits)).

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release.
- [x] Create (or move) `v1` tag that points to new release.

felddy committed:
Add cyhy_db package and models

Delete example package and files

Update Python version to 3.12 in build workflows

Update package name and description in setup.py

Update dependencies in setup.py

felddy committed:
Apply changes made by linters

Add comments to severity calculation.

felddy committed:
Add doc strings where needed

felddy committed:
Simplify field validation

felddy committed:
Add test case for invalid CVSS score

felddy committed:
Switch to MongoEngine

felddy committed:
Add utcnow function to utils module

felddy committed:
Update MongoDB connection and add support for Mongo Express

felddy committed:
Refactor CVE model and test

This commit refactors the CVE model and test in the `cyhy_db` package. The changes include:
- Replacing the `odmantic` library with `mongoengine` for the model definition.
- Updating the field types and constraints in the CVE model.
- Removing the `calculate_severity` method and integrating the severity calculation directly in the `save` method.
- Updating the test cases to reflect the changes in the model.

felddy committed:
Add IPAddressField and test cases for IP address validation and conversion

felddy committed:
Add ScanDoc and SnapshotDoc models

felddy committed:
Sort imports

felddy committed:
Implement CVE, ScanDoc with beanie

felddy committed:
Add cyhy_db package and initialize_db function

felddy committed:
Update conftest.py with Docker and mongo express fixtures

felddy committed:
Add test for database connection using motor and beanie libraries

felddy committed:
Refactor imports and update test cases in test_cve.py

felddy committed:
Remove test_ip_address.py

felddy committed:
Add tests for ScanDoc model functionality

felddy committed:
Add asyncio_mode to pytest.ini

felddy committed:
Switch to beanie ORM as a dependency

felddy committed:
Add enum classes for various types and statuses

felddy committed:
Add RequestDoc model and test

felddy committed:
Refactor CVE model and add model validator for calculating severity

felddy committed:
Rename ScanDoc test

felddy committed:
Add RequestDoc model to ALL_MODELS list

felddy committed:
Add new dependencies to setup.py

felddy committed:
Add test case to reset latest flag by owner

felddy committed:
Sort imports

felddy committed:
Refactor ScanDoc model and add new tests

felddy committed:
Add fields and indexes to SnapshotDoc model

felddy committed:
Organize imports

felddy committed:
Add HostDoc model and test cases

felddy committed:
Update imports and type annotations in scan_doc.py

felddy committed:
Add new models and fields to SnapshotDoc

felddy committed:
Reorder model imports to avoid circular references

felddy committed:
Add Protocol enum for TCP and UDP

felddy committed:
Add deprecated decorator to utils

felddy committed:
Refactor and partially implement helpers with deprecation

felddy committed:
Add HostScanDoc model with settings

felddy committed:
Add KEVDoc model

felddy committed:
Add PlaceDoc model

felddy committed:
Add PortScanDoc model with indexes

felddy committed:
Add ReportDoc model

felddy committed:
Replace deprecated datetime utcnow calls

felddy committed:
Add SystemControlDoc model

felddy committed:
Add TallyDoc model

felddy committed:
Add VulnScanDoc model

felddy committed:
Update import statements in db.py

felddy committed:
Replace deprecated utcnow calls

felddy committed:
Refactor test_find to test_get_by_ip

felddy committed:
Add test data generator for CVE and RequestDoc models

felddy committed:
Add .hypothesis and .vscode to .gitignore

felddy committed:
Add note about pydantic extras

felddy committed:
Add NotificationDoc model to cyhy_db

felddy committed:
Update model configurations to use ConfigDict instead of Config class

felddy committed:
Organize imports per isort

felddy committed:
Add import statement for SnapshotDoc in ScanDoc

felddy committed:
Apply changes made by linters

felddy committed:
Remove unused index in RequestDoc model

felddy committed:
Add module docstring to disable isort breaking import order.

felddy committed:
Clean up unused imports

felddy committed:
Add TODO to delete BASESCAN if unused

felddy committed:
Improve comments on utcnow util

felddy committed:
Replace date factory with custom function.

felddy committed:
Add ScanDoc model header comment

felddy committed:
Add TODO to inherit from BaseModel

I can't remember why I added this TODO.  Someday I will, and that day will be glorious.

felddy committed:
Add non-interactive flag to mypy types install

The script will pause and ask for user input without this flag.

felddy committed:
Bump docker library to fix upstream break by requests

See:
- https://github.com/docker/docker-py/pull/3257

felddy committed:
Add types and ignores to make mypy happy

felddy committed:
Bump dependency versions to latest

dav3r committed:
Resolve some flake8 issues

* Remove unused imports
* Specify all imports, instead of using wildcard imports
* Use preferred comparisons
* Adhere to docstring whitespace and imperative mood rules

dav3r committed:
Satisfy isort and black linters

felddy committed:
Update comment

felddy committed:
Add py.typed file to allow users to type check

dav3r committed:
Ignore some E712 flake8 findings

Beanie syntax requires us to write code in a way that triggers E712.

dav3r committed:
Appease flake8 by using an otherwise unused variable

Error was: F841 local variable 'host_doc' is assigned to but never used

dav3r committed:
Add a whole bunch of docstrings to placate flake8

dav3r committed:
Tell bandit to ignore warnings about our use of random numbers

We aren't using Random() for the purposes of cryptography here, so we can safely ignore these warnings.

felddy committed:
Limit Python version to 3.10 and greater

We are using the or (`|`) operator for types that was introduced in Python 3.10 with PEP 604.

See:
- https://docs.python.org/3/whatsnew/3.10.html#pep-604-new-type-union-operator

felddy committed:
Uncomment dependabot ignores handled by skeleton

dav3r committed:
Explicitly set the pytest log level to INFO

This is primarily because pymongo DEBUG level logging is extremely verbose and should only be enabled when needed.

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Restore a useful comment and import from the skeleton

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Remove a commented out line that is no longer needed

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Add two unit tests for the initialize_db function

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Correctly set state in HostDoc

dav3r committed:
Add unit tests for HostDoc.set_state()

dav3r committed:
Add unit tests for SystemControlDoc

dav3r committed:
Add additional unit tests for ScanDoc.tag_latest()

dav3r committed:
Add a "no op" unit test for HostDoc.set_state()

dav3r committed:
Add unit tests for utils/decorators.py

dav3r committed:
Improve field validator for start window in RequestDoc

This field now accepts either a string representing the time (%H:%M:%S) or a datetime.time object, which is stored in the database as a string in %H:%M:%S format.

This also improves our input checking so that invalid times (e.g. 34:45:56) can no longer be added.

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Add unit tests for parse_time in Window (request_doc.py)

Also, remove some commented-out code that is no longer needed.

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Avoid using the reserved "_id" field name in TallyDoc

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Add unit tests for TallyDoc

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Output mongodb URI at end of pytest run when used with --mongo-express option

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Update README from skeleton to include useful information

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Remove a TODO that is no longer relevant

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Make all children of ScanDoc (HostScanDoc, PortScanDoc, VulnScanDoc) include the same indexes as ScanDoc

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Update a comment

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

felddy committed:
Remove duplicate note text

felddy committed:
Make ScanDoc abstract and document

felddy committed:
Add stub of TicketDoc model

felddy committed:
Improve comment for protocol enum

felddy committed:
Rename TicketEvent

felddy committed:
Add custom CyHy exceptions

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Add TIcketDoc implementation

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

dav3r committed:
Use correct syntax when appending a TicketDoc event

dav3r committed:
Remove an unnecessary async declaration

dav3r committed:
Correctly define TicketEvent delta to default to None

Since delta is optional, it makes sense to default to None.

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Prefer serialization_alias to alias

For more info, see https://docs.pydantic.dev/latest/concepts/fields/#field-aliases

Co-authored-by: Mark Feldhousen <mark.feldhousen@gwe.cisa.dhs.gov>

dav3r committed:
Use correct query syntax

dav3r committed:
The "In" query operator requires an array

dav3r committed:
Correct issues with latest_port and latest_vuln in TicketDoc

dav3r committed:
Make TicketEvent.reference optional, defaulting to None

Some event types (e.g. CHANGED) do not require a reference.

dav3r committed:
Add unit tests for TicketDoc

dav3r committed:
Make TicketDoc.add_event() raise the same exception in all currently-supported Python versions

dav3r committed:
Remove commented-out legacy functions

There are a lot of functions in the Python 2 version of HostDoc (such as the ones being deleted here) that may or may not be used.  Instead of porting them all over to cyhy-db, we plan to port them as they are needed and rewrite them to take advantage of Python 3.

felddy committed:
Convert enums to StrEnum and use auto()

This requires a minimum Python version of 3.11

felddy committed:
Remove unused BASESCAN enum value in Stage

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Reverting change to CVSSVersion enum

felddy committed:
Convert test to use CVSS enumeration

felddy committed:
Rename CVE to CVEDoc

This change normalizes the names of all
document classes.

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

jsf9k commented on .github/dependabot.yml:
```diff
@@ -19,10 +19,10 @@ updates:
       - dependency-name: hashicorp/setup-terraform
       - dependency-name: mxschmitt/action-tmate
       - dependency-name: step-security/harden-runner
-      # # Managed by cisagov/cyhy-db
```
This is an indication that you need to run `gh extension upgrade --all` locally.

dav3r replied:
I think @felddy did that at some point after he created this new repo, but it will be good for him to confirm.

jsf9k replied:
It was definitely @felddy [who made the change](https://github.com/cisagov/cyhy-db/blame/first-commits/.github/workflows/build.yml#L65).  Maybe the version wasn't 3.12 yet then.  Soon it will be 3.13, which is already out.

In any event, the change can and should be reverted now.

felddy replied:
Updated `gh` extension.  Do we need a change here?

jsf9k replied:
Nope.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -248,7 +244,7 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: ${{ steps.setup-env.outputs.python-version }}
+          python-version: "3.12"
```
The version from cisagov/setup-env-github-action [is already 3.12](https://github.com/cisagov/setup-env-github-action/blob/develop/action.yml).  Why have you hardcoded this line?

```suggestion
          python-version: ${{ steps.setup-env.outputs.python-version }}
```

felddy replied:
See:
- 0f917aeb84c953c25cf412a0eabce627aca6f7b3

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -62,7 +62,7 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: ${{ steps.setup-env.outputs.python-version }}
+          python-version: "3.12"
```
The version from cisagov/setup-env-github-action [is already 3.12](https://github.com/cisagov/setup-env-github-action/blob/develop/action.yml).  Why have you hardcoded this line?

```suggestion
          python-version: ${{ steps.setup-env.outputs.python-version }}
```

felddy replied:
I don't think it was `3.12` when I cloned.  🤷 (maybe)

In any case.  Fixed in:
- 0f917aeb84c953c25cf412a0eabce627aca6f7b3

jsf9k commented on setup.py:
```diff
@@ -102,11 +105,15 @@ def get_version(version_file):
             # 1.11.1 fixed this issue, but to ensure expected behavior we'll pin
             # to never grab the regression version.
             "coveralls != 1.11.0",
+            "docker == 7.1.0",
```
Why does this one test dependency have a hard-coded version?  If a particular version is required then there should be a comment explaining why.

felddy replied:
Fixed. 
See:
- 8fff2b4a827029cb75dee35f4f4ebe3b1b47ad30

jsf9k replied:
See commit 8fff2b4.

jsf9k reviewed (changes requested):
In addition to the quibbles below, when I run `./setup-env -p 3.13.0 -f` I get an error:
```console
src/cyhy_db/models/tally_doc.py:41: error: Incompatible types in assignment (expression has type "str", base class "Document" defined the type as "PydanticObjectId | None")  [assignment]
```

I did confirm that all tests pass by running `pytest -vs --mongo-express`.

dav3r committed:
Sort attributes in all document classes

dav3r commented on src/cyhy_db/models/cve_doc.py:
```diff
@@ -0,0 +1,51 @@
+    @model_validator(mode="before")
+    def calculate_severity(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Calculate CVE severity based on the CVSS score and version."""
+        if values["cvss_version"] == "2.0":
```
I think we missed this one:
```suggestion
        if values["cvss_version"] == CVSSVersion.V2:
```

felddy replied:
Change applied: 
- e5fdec21e6b9dbb9d44544c63dd5cc64d1bf4d7e

felddy committed:
Clean up dependencies

- Remove unnecessary pins.
- Remove unused dependencies.
- `coveralls` comment is no longer required.
- `setuptools` pin and comment is no longer required.

See:
- https://github.com/cisagov/cyhy-config/commit/8df2e692d2920743c40368a496ca154769b2985a

felddy committed:
Replace magic string with proper enum

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Revert hard coding of Python version

felddy committed:
Work around type assignment error

Add a mypy ignore directive to TallyDoc.
Document all uses of this directive in a new issue.

felddy commented:
> In addition to the quibbles below, when I run `./setup-env -p 3.13.0 -f` I get an error:
> 
> ```
> src/cyhy_db/models/tally_doc.py:41: error: Incompatible types in assignment (expression has type "str", base class "Document" defined the type as "PydanticObjectId | None")  [assignment]
> ```

I was able to replicate this.  I've applied the same "fix" to `TallyDoc` that we used in the other model `id` fields.  I've opened a new issue to track why this is the case, and added comments pointing to the issue.

See:
- fd82fe948ca88fe2e8f949842987468b3472d564
- https://github.com/cisagov/cyhy-db/issues/7

dav3r commented on src/cyhy_db/models/place_doc.py:
```diff
@@ -0,0 +1,38 @@
+
+    model_config = ConfigDict(extra="forbid")
+
+    clazz: str = Field(alias="class")  # 'class' is a reserved keyword in Python
```
We should be consistent with how we handle these cases (e.g. `class_` -> `class`).  We also need to nail down the "alias" vs. "serialization_alias" business.  For reference: https://github.com/cisagov/cyhy-db/blob/5b2bb40c9984243cc01a870175ff07ad05167925/src/cyhy_db/models/ticket_doc.py#L30-L33

jsf9k replied:
I think `clazz` is a pretty common alternate name.  It's used a lot in Java, e.g., where the need to name things `class` comes up surprisingly often.

Also, I like to think of `clazz` as being a classy guy _but with an edge_.  Like a James Bond sort of character.

dav3r replied:
I don't have anything against `clazz`, but if `class` is aliased as `clazz`, then should `from` be aliased as `fromm`, `fromz`, `frozz`, or something else?

It just seems like appending an underscore to create the alias is a simpler way to go:
`class` -> `class_`
`from` -> `from_`

felddy replied:
This most certainly was my Java neurons solving a reserved word problem.  I agree that it should be normalized with our previous usage of `from_`.  Although `fromz` is tempting.

felddy replied:
Resolved in:
- 07724b23d80129fea518e373f5825469882bc800

dav3r replied:
Did you also confirm if we want `alias` or `serialization_alias`, [as mentioned above](https://github.com/cisagov/cyhy-db/pull/5#discussion_r1801711616)?

felddy replied:
I think I've tracked it down.  `alias` should be all we need but I see evidence of either a bug, a lack of documentation, or my own misunderstanding of how all this works.

I think it has to do with how aliases are being applied to sub-documents.  This leads me to believe that it is a bug: 
- https://github.com/BeanieODM/beanie/issues/369

But the `populate_by_name` model configuration option of Pydantic leads me to believe it might be by design...
- https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name

In any case, if we just set up our alias like this it fails with a Pydantic validation error:
```python
    from_: Optional[bool | float | int | str] = Field(..., alias="from")
```

But if I set `populate_by_name` to `true` it works as expected: 
```python
class EventDelta(BaseModel):
    """The event delta model."""

    model_config = ConfigDict(populate_by_name=True)

    from_: Optional[bool | float | int | str] = Field(..., alias="from")
```

So `populate_by_name` needs to be `true`, as either the proper way to do this, or as a work-around to a bug.

dav3r replied:
Excellent detective work, thanks for running this all to ground! 🚀

dav3r commented on src/cyhy_db/models/request_doc.py:
```diff
@@ -0,0 +1,134 @@
+
+    model_config = ConfigDict(extra="forbid")
+
+    scan_type: ScanType = Field(..., alias="scanType")
```
Why is this alias here?

felddy replied:
Resolved:
- 438c199f0ecd5616c3e7a444d772bc81f6a1ee0c

dav3r replied:
That works, thanks!

dav3r commented on src/cyhy_db/models/scan_doc.py:
```diff
@@ -0,0 +1,120 @@
+
+    @classmethod
+    async def tag_latest(
+        cls, owners: List[str], snapshot: Union["SnapshotDoc", ObjectId, str]
```
Remove unneeded quotes:
```suggestion
        cls, owners: List[str], snapshot: Union[SnapshotDoc, ObjectId, str]
```

jsf9k commented:
> > In addition to the quibbles below, when I run `./setup-env -p 3.13.0 -f` I get an error:
> > ```
> > src/cyhy_db/models/tally_doc.py:41: error: Incompatible types in assignment (expression has type "str", base class "Document" defined the type as "PydanticObjectId | None")  [assignment]
> > ```
> 
> I was able to replicate this. I've applied the same "fix" to `TallyDoc` that we used in the other model `id` fields. I've opened a new issue to track why this is the case, and added comments pointing to the issue.
> 
> See:
> 
> * [5b2bb40](https://github.com/cisagov/cyhy-db/commit/5b2bb40c9984243cc01a870175ff07ad05167925)
> * [Revisit custom `id` field types and `type: ignore[assignment]` #7](https://github.com/cisagov/cyhy-db/issues/7)

Thank you!  I was able to verify that I am now able to run `./setup-env -p 3.13.0 -f` with no errors.

felddy committed:
Work around type assignment error

Add a mypy ignore directive to TallyDoc.
Document all uses of this directive in a new issue.

dav3r reviewed (approved):
Strong work! 💪 💼 

Please take a look at the couple of minor thangs I noticed.

dav3r committed:
Merge branch 'first-commits' of github.com:cisagov/cyhy-db into first-commits

felddy committed:
Rename reserved word deconfliction to match previously used pattern

The reserved word `from` was replaced with `from_` in `TicketDoc`.  This follows that pattern.

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>
Co-authored-by: Jeremy Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

felddy committed:
Remove field alias

This breaking change is documented in:
- Data migration from `cyhy-core` database #8

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

felddy committed:
Remove inexplicable quoting

😜

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

felddy committed:
Add asyncio_default_fixture_loop_scope configuration option to pytest.ini

felddy committed:
Enable populate_by_name to handle aliases

I'm not sure if this is a work-around for a bug, or a needed option in Pydantic.

See:
- https://github.com/BeanieODM/beanie/issues/369
- https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name

felddy committed:
Add ScanLimit save test

felddy committed:
Attempt to improve ScanDoc test coverage

In an attempt to reach 100% coverage I broke out the various ways `reset_latestest_by_ip` could be called... to no avail.

I believe this issue explains why the list comprehension shows a branch part as uncovered:
- https://github.com/nedbat/coveragepy/issues/1617

felddy committed:
Sort pytest options

dav3r reviewed (approved):
👍 👍

felddy committed:
Update version file path in bump_version.sh

felddy committed:
Bump version from 0.0.1 to 1.0.0
